http-body-util/src/channel.rs

```diff
@@ -0,0 +1,159 @@
+//! A body backed by a channel.
+
+use std::{
+    fmt::Display,
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use bytes::Buf;
+use http::HeaderMap;
+use http_body::{Body, Frame};
+use tokio::sync::mpsc;
+
+/// A body backed by a channel.
+pub struct Channel<D, E = std::convert::Infallible> {
+    rx_frame: mpsc::Receiver<Frame<D>>,
+    rx_error: mpsc::Receiver<E>,
```

Comment:
Would a `oneshot` channel work? It should be a smaller struct, and use less memory wen sending the error (the mpsc channel reserves blocks).

Reply:
At first I thought it wouldn't be possible because `one shot::Receiver` doesn't have a `poll_recv` method but I suppose I can call `Future::poll` instead 🤔 I'll try that.

Reply:
i ran into this issue myself, and noticed this thread here. i've opened tokio-rs/tokio#7059 to introduce a `poll_recv(..)` method for `oneshot::Receiver<T>`.